File: clients/kotlin/src/main/kotlin/play/loki/sdk/LokiQuantize.kt
package play.loki.sdk

import kotlin.math.round

object LokiQuantize {
    fun quantize(value: Double, scale: Long): Long {
        require(value.isFinite()) { "quantize requires a finite value" }
        require(scale > 0) { "quantize requires a positive safe-integer scale" }
        val quantized = round(value * scale)
        require(quantized.isFinite() && quantized >= Long.MIN_VALUE.toDouble() && quantized <= Long.MAX_VALUE.toDouble()) {
            "quantized value is not a finite safe integer"
        }
        return quantized.toLong()
    }

    fun dequantize(value: Long, scale: Long): Double {
        require(scale > 0) { "dequantize requires a positive safe-integer scale" }
        return value.toDouble() / scale
    }
}

File: clients/unity/Runtime/LokiQuantize.cs
using System;

namespace Loki.Play.SDK
{
    public static class LokiQuantize
    {
        public static long Quantize(double value, long scale)
        {
            if (double.IsNaN(value) || double.IsInfinity(value))
                throw new ArgumentOutOfRangeException(nameof(value), "quantize requires a finite value");
            if (scale <= 0)
                throw new ArgumentOutOfRangeException(nameof(scale), "quantize requires a positive safe-integer scale");
            var quantized = Math.Round(value * scale);
            if (quantized < long.MinValue || quantized > long.MaxValue)
                throw new OverflowException("quantized value is not a finite safe integer");
            return (long)quantized;
        }

        public static double Dequantize(long value, long scale)
        {
            if (scale <= 0)
                throw new ArgumentOutOfRangeException(nameof(scale), "dequantize requires a positive safe-integer scale");
            return (double)value / scale;
        }
    }
}
